### PR TITLE
fix(ci): harden PR-size exception governance

### PR DESCRIPTION
**Pull request by qnbs**
## Summary

- Enforce the ordinary absolute PR-size ceiling (30 meaningful files, 3000 meaningful lines, 15 commits) on every active exception registry entry.
- Keep exception authority base-ref governed and prevent governed PRs from authorizing changes to the registry, checker, declaration, or CI workflow.
- Require short single-line exception rationales and retire the historical PR #596 record without changing its recorded authorization values.
- Add focused regression coverage for absolute ceilings, allowlists, self-expansion, historical entries, and valid boundary/supplemental exceptions.

## Why

Issue #625 hardens the exception mechanism against self-expansion and retroactive elevation. This PR does not redesign the trusted workflow boundary owned by #510.

## Validation

- `pnpm exec vitest run tests/unit/tooling/checkPrSize.test.ts` — 68 passed
- `pnpm run ci:prepush` — passed
- `git diff --check` — passed
- targeted Biome — passed
- signed commit verified locally

## Scope

No runtime product behavior, #553 filesystem semantics, or #654 source was changed. The README test-count update is the mechanically required docs-truth correction reported by `ci:prepush`.

## Summary by Sourcery

Harden pull-request size exception governance by enforcing absolute limits and preventing exceptions from modifying the controls that govern them.

New Features:
- Add governance safeguards that constrain and validate pull-request size exceptions.

Bug Fixes:
- Prevent active exceptions from exceeding the ordinary 30-file, 3,000-line, and 15-commit ceilings or authorizing changes to governance-control files.
- Prevent historical exceptions from being applied to current pull requests.

Enhancements:
- Require unique, safe exception paths, recognized non-executable supplemental artifacts, and short single-line rationales.
- Preserve historical exception records while keeping their legacy authorization values unchanged.

Documentation:
- Update README test-count metrics to reflect the expanded test suite.

Tests:
- Add regression coverage for governance bypasses, absolute limits, historical exceptions, supplemental allowances, and boundary cases.

___

## **CodeAnt-AI Description**
Prevent PR-size exceptions from weakening the rules that govern them

### What Changed
- PR-size exceptions can no longer authorize changes to the exception registry, checker, type declarations, or CI workflow.
- Active exceptions must stay within the ordinary limits of 30 files, 3,000 meaningful lines, and 15 commits; invalid entries are rejected.
- Historical exceptions remain recorded but are no longer applied to current pull requests.
- Exception reasons are limited to short, single-line text, with validation for duplicate and unsafe paths.
- Added regression coverage for governance bypasses, absolute limits, historical entries, and valid boundary cases.
- Updated documented test counts to match the current suite.

### Impact
`✅ Fewer PR-size governance bypasses`
`✅ Consistent enforcement of 30-file, 3,000-line, and 15-commit ceilings`
`✅ Historical exceptions no longer affect active pull requests`
<details><summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

---
## Summary by cubic
Harden PR-size exception governance per issue #625: active exceptions are now capped at the absolute ceiling (30 files, 3000 meaningful lines, 15 commits) and can no longer let a governed PR rewrite the files that govern it.

**Behavior changes**
- Exception entries that exceed the absolute ceiling now fail closed instead of being applied.
- Governance control paths (`config/pr-size-exceptions.json`, `scripts/ci-prepush-classifier.mjs`, `scripts/check-pr-size.mjs`, `scripts/check-pr-size.d.mts`, `.github/workflows/ci.yml`) are rejected from `allowedPaths`, and any PR touching them cannot use an exception.
- Supplemental line allowances are only allowed for recognized non-executable artifacts (docs, generated reports) and must already be in `allowedPaths`.
- Exception rationales must now be a single line of at most 500 characters.
- The PR #596 record is marked `status: "historical"`, preserving its recorded values without ever applying them.
- Added regression coverage for absolute ceilings, allowlists, self-expansion, historical entries, and boundary/supplemental cases; README test count updated to match (7516+).

<sup>Written for commit d708ce4fb9639c17a8b0c1f4bfe706dea7a41619. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/qnbs/WorldScript-Studio/pull/657?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

- **Documentation**
  - Updated project metrics to reflect 7,516+ tests across 602 files.

- **Chores**
  - Improved pull request size validation with clearer exception handling and stricter limits.
  - Added protections preventing size exceptions from covering governance-control changes.
  - Historical exceptions are now recorded separately and cannot be used for new resolutions.
  - Expanded validation for exception status, limits, allowed paths, justification format, and supplemental allowances.
  - Added broader test coverage for boundary conditions, registry validation, and historical exceptions.

### CODE REVIEW

File: scripts/check-pr-size.mjs [hunk 1]
```diff
@@ -162,6 +169,116 @@ function isSafeRegistryPath(value) {
...
+      `invalid ${EXCEPTION_REGISTRY_PATH}: exception ${index} needs supplementalLineAllowances`,
+    );
+  }
+  const supplementalPaths = new Set();
+  for (const allowance of entry.supplementalLineAllowances) {
+    if (
+      !allowance ||
+      !isSafeRegistryPath(allowance.path) ||
+      !isValidPositiveInteger(allowance.maxMeaningfulLines)
+    ) {
```
[the remark this answers is not in the document]
Comment (qnbs) [reply scripts/check-pr-size.mjs:242]: Fixed in d708ce4fb9639c17a8b0c1f4bfe706dea7a41619. Active supplementalLineAllowances now fail closed unless their exact paths match recognized non-executable document/generated/report artifact patterns; executable paths such as services/security.ts remain counted. Added the regression; Biome, git diff --check, 74/74 targeted tests, and ci:prepush pass.

File: scripts/check-pr-size.mjs [hunk 2]
```diff
@@ -14,6 +14,13 @@ const TIERS = {
 
 const EXCEPTION_REGISTRY_PATH = 'config/pr-size-exceptions.json';
 const REQUIRED_EXCEPTION_SCHEMA_VERSION = 1;
+const MAX_EXCEPTION_REASON_LENGTH = 500;
+const GOVERNANCE_CONTROL_PATHS = new Set([
+  EXCEPTION_REGISTRY_PATH,
+  '.github/workflows/ci.yml',
+  'scripts/check-pr-size.d.mts',
+  'scripts/check-pr-size.mjs',
```
[the remark this answers is not in the document]
Comment (qnbs) [reply scripts/check-pr-size.mjs:23]: Fixed in d708ce4fb9639c17a8b0c1f4bfe706dea7a41619. scripts/ci-prepush-classifier.mjs is now a protected governance-control path, and both changed-path and base-registry-allowlist regressions cover it. The correction was signed and pushed normally.
